neural_lam/weather_dataset.py

```diff
@@ -294,46 +444,133 @@ def _slice_forcing_time(self, da_forcing, idx, n_steps: int):
-                da_sliced = da_sliced.assign_coords(
-                    window=np.arange(len(da_sliced.window))
+                # Assign 'window' coordinate
+                da_window = da_window.assign_coords(
+                    window=np.arange(-num_past_steps, num_future_steps + 1)
                 )
 
-                # Add a 'time' dimension to keep track of steps using actual
-                # time coordinates
-                current_time = da_forcing.time[offset + step]
-                da_sliced = da_sliced.expand_dims(
-                    dim={"time": [current_time.values]}
-                )
+                da_window = da_window.expand_dims(dim={"time": [state_time]})
 
-                da_list.append(da_sliced)
+                da_list.append(da_window)
+
+        da_forcing_matched = xr.concat(da_list, dim="time")
+
+        # Generate temporal embedding `time_diff_steps` for the
+        # forcing/boundary data. This is the time difference in multiples
+        # of state time steps between the forcing/boundary time and the
+        # state time
 
-            # Concatenate the list of DataArrays along the 'time' dimension
-            da_concat = xr.concat(da_list, dim="time")
+        if is_boundary:
+            if self.datastore_boundary.is_forecast:
+                boundary_time_step = self.forecast_step_boundary
+                state_time_step = self.forecast_step_state
+            else:
+                boundary_time_step = self.time_step_boundary
+                state_time_step = self.time_step_state
+            time_diff_steps = (
+                da_forcing_matched["window"]
+                * (boundary_time_step / state_time_step),
+            )
+        else:
+            if self.datastore.is_forecast:
+                forcing_time_step = self.forecast_step_forcing
+                state_time_step = self.forecast_step_state
+            else:
+                forcing_time_step = self.time_step_forcing
+                state_time_step = self.time_step_state
+            time_diff_steps = (
+                da_forcing_matched["window"]
+                * (forcing_time_step / state_time_step),
+            )
+        time_diff_steps = da_forcing_matched.isel(
+            grid_index=0, forcing_feature=0
+        ).window.values
+        # Add time difference as a new coordinate to concatenate to the
+        # forcing features later as temporal embedding
+        da_forcing_matched["time_diff_steps"] = (
+            ("window"),
+            time_diff_steps,
+        )
 
-        return da_concat
+        return da_state_sliced, da_forcing_matched
+
+    def _process_windowed_data(self, da_windowed, da_state, da_target_times):
+        """Helper function to process windowed data. This function stacks the
+        'forcing_feature' and 'window' dimensions and adds the time step
+        differences to the existing features as a temporal embedding.
```

Comment:
I think this is highly related. The temporal embedding (which we should probably not refer to as such, as this is not an embedding, so from here on: the delta times) is the time differences for the different entries in the window. After these have been stacked it is not clear to me how you would know what the delta-times are, since you've got rid of the window dimension.